### ISSUE TITLE
Error for `useStore`: Warning: The result of getSnapshot should be cached to avoid an infinite loop

### ISSUE THREAD

**Issue by adamtaylor13** [nanostores/react#22]
I'm using nanostores (`0.9.3`), in astro (`1.9.0`) with react (`18.0.0`).

I've got this store:
```ts
import { DataResponse } from "@components/hooks/useApi";
import { createFetcherStore } from "@stores/apiFetcher";
import { useStore } from "@nanostores/react";

export const $account = createFetcherStore<DataResponse<"/account">>([
    "/account",
]);

export const useAccount = () => {
    const fetchResults = useStore($account);
    
    // ...

    return fetchResults;
};
```

Used in a React component like so:
```tsx
export default function NavAuthOptions() {
    const { data: account } = useAccount();
    // ...
}
```

And I get this error in the console:
![CleanShot 2023-09-13 at 09 51 56@2x](https://github.com/nanostores/react/assets/2922753/d570e766-175e-4505-b014-0aa0d076988a)

I should note however that there doesn't actually appear to be any errors. The site seems to function just fine, which may actually be more disconcerting than the error itself!

My Google-Fu has failed me here as the only semi-related issue I could find was this https://github.com/pmndrs/zustand/discussions/1936, however it doesn't appear related as I don't have a `.getState()` function.

I've also tried just using `useStore($account)` directly inside my component instead of importing `useAccount`, but the error still shows in the console.

**Comment by grctest** [comment]
Also getting this error using nanostore query/react with react + astro

Any luck resolving it?

**Comment by adamtaylor13** [comment]
@grctest Unfortunately no—we've just shipped it with the error, which hasn't seemed to affect our site at all. But like I said above, that's more disconcerting than the error itself. Why would we be logging errors if there isn't actually an error? So far no one from nanostores has reached out to assist with this :/

**Comment by grctest** [comment]
It seems like it's a known React issue? https://github.com/facebook/react/issues/24508#issuecomment-1121380340

**Comment by constantineMelios85** [comment]
Is there any updates on this issue. I have the same with nanostores/query and nanostores/react.

**Comment by ai** [comment]
I am not using React anymore, so someone else needs to do research and find a solution.

**Comment by JosefJezek** [comment]
I found this solution...
![image](https://github.com/user-attachments/assets/97fb1b3b-d58e-4576-bf6e-59475214bf38)

useStoreData() does call render second time :-/
![image](https://github.com/user-attachments/assets/631774a2-d883-428c-93f3-88d773b26c96)

**Comment by JosefJezek** [comment]
@ai could you fix this issue?

**Comment by ai** [comment]
@JosefJezek can use send PR? I can’t copy code from screenshots.

**Comment by JosefJezek** [comment]
my change in useStore...
```js
const getKey = () => {
  const store = get();
  return storeKey ? store[storeKey] : store;
};

return useSyncExternalStore(subscribe, getKey, getKey);
```

**Comment by ai** [comment]
@JosefJezek we don’t have `storeKey` in the `useStore`. Please, send PR, it will be faster.

**Comment by JosefJezek** [comment]
my solution is not perfect... we need `const { data, loading, error } = useStore($post);`, not only one key `data`

please fix `createFetcherStore`, this store has too many changes

**Comment by yuyi919** [comment]
I've encountered an same issue where `listenKeys` in `createFetcherStore` causes excessive responses (or updates). To address this, I implemented the following temporary solution, which works for my use case:

```typescript
export function useStoreFixed<SomeStore extends Store>(
  store: SomeStore,
  opts: UseStoreOptions<SomeStore> = {},
): StoreValue<SomeStore> {
  const snapshot = useRef(store.get());
  const get = () => snapshot.current;
  return useSyncExternalStore(
    useCallback(
      (onChange) =>
        opts.keys
          ? listenKeys(store as any, opts.keys, (changed) => {
              const temp = snapshot.current;
              if (changed === temp) return;
              if (changed == null || temp == null) return;
              if (
                opts.keys!.some((key) => changed[key] !== snapshot.current[key])
              ) {
                snapshot.current = changed;
                onChange();
              }
            })
          : store.listen((changed) => {
              snapshot.current = changed;
              onChange();
            }),
      [opts.keys, store],
    ),
    get,
    get,
  );
}
```
Ideally, this issue should be resolved on the library's end to prevent unnecessary reactivity.

**Comment by yuyi919** [comment]
I've encountered an same issue where `listenKeys` in `createFetcherStore` causes excessive responses (or updates). To address this, I implemented the following temporary solution, which works for my use case:

```typescript
export function useStoreFixed<SomeStore extends Store>(
  store: SomeStore,
  opts: UseStoreOptions<SomeStore> = {},
): StoreValue<SomeStore> {
  const snapshot = useRef(store.get());
  const get = () => store.get();
  return useSyncExternalStore(
    useCallback(
      (onChange) =>
        opts.keys
          ? listenKeys(store as any, opts.keys, (changed) => {
              const temp = snapshot.current;
              if (
                changed !== temp &&
                (changed == null ||
                  temp == null ||
                  opts.keys!.some(
                    (key) => changed[key] !== snapshot.current[key],
                  ))
              ) {
                snapshot.current = changed;
                onChange();
              }
            })
          : store.listen((changed) => {
              onChange();
            }),
      [opts.keys, store],
    ),
    get,
    get,
  );
}
```
Ideally, this issue should be resolved on the library's end to prevent unnecessary reactivity.

**Comment by ai** [comment]
The fix by @yuyi919 was released in 0.8.0.
